lib/spack/spack/package_base.py

```diff
@@ -1187,26 +1186,12 @@ def installed_upstream(self):
         warnings.warn(msg)
         return self.spec.installed_upstream
 
-    def _make_fetcher(self):
-        # Construct a composite fetcher that always contains at least
-        # one element (the root package). In case there are resources
-        # associated with the package, append their fetcher to the
-        # composite.
-        root_fetcher = fs.for_package_version(self)
-        fetcher = fs.FetchStrategyComposite()  # Composite fetcher
-        fetcher.append(root_fetcher)  # Root fetcher is always present
-        resources = self._get_needed_resources()
-        for resource in resources:
-            fetcher.append(resource.fetcher)
-        fetcher.set_package(self)
-        return fetcher
```

Comment:
We have unit tests for this, but we cannot simply use `self.fetcher` :slightly_frowning_face: [^1] Due to early design decisions, a `fetcher` gets a `url` on construction - so we need one for each `patch` or resource we want to fetch.

As an example, trying to:
```console
$ spack mirror create -d /tmp/fetcher aomp
```
both on this PR and on `develop`, we obtain in the PR:
```
PR/
├── aomp
│   ├── amd-llvm-project-3.10.0.tar.gz -> ../_source-cache/archive/ce/ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
│   ├── aomp-3.10.0.tar.gz -> ../_source-cache/archive/ce/ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
│   ├── aomp-extras-3.10.0.tar.gz -> ../_source-cache/archive/ce/ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
│   ├── flang-3.10.0.tar.gz -> ../_source-cache/archive/ce/ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
│   └── rocm-device-libs-3.10.0.tar.gz -> ../_source-cache/archive/ce/ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
└── _source-cache
    └── archive
        └── ce
            └── ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
```
while on develop we have:
```
develop/
├── aomp
│   ├── amd-llvm-project-3.10.0.tar.gz -> ../_source-cache/archive/3f/3ff18a8bd31d5b55232327e574dfa3556cf26787e105d0ba99411c5687325a8d.tar.gz
│   ├── aomp-3.10.0.tar.gz -> ../_source-cache/archive/ce/ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
│   ├── aomp-extras-3.10.0.tar.gz -> ../_source-cache/archive/ee/ee146cff4b9ee7aae90d7bb1d6b4957839232be0e7dab1865e0ae39832f8f795.tar.gz
│   ├── flang-3.10.0.tar.gz -> ../_source-cache/archive/81/81674bf3c9d8fd9b16fb3e5c66a870537c25ff8302fc1b162ab9e95944167163.tar.gz
│   └── rocm-device-libs-3.10.0.tar.gz -> ../_source-cache/archive/bc/bca9291385d6bdc91a8b39a46f0fd816157d38abb1725ff5222e6a0daa0834cc.tar.gz
└── _source-cache
    └── archive
        ├── 3f
        │   └── 3ff18a8bd31d5b55232327e574dfa3556cf26787e105d0ba99411c5687325a8d.tar.gz
        ├── 81
        │   └── 81674bf3c9d8fd9b16fb3e5c66a870537c25ff8302fc1b162ab9e95944167163.tar.gz
        ├── bc
        │   └── bca9291385d6bdc91a8b39a46f0fd816157d38abb1725ff5222e6a0daa0834cc.tar.gz
        ├── ce
        │   └── ce29cead5391a4a13f2c567e2e059de9291888d24985460725e43a91b740be7a.tar.gz
        └── ee
            └── ee146cff4b9ee7aae90d7bb1d6b4957839232be0e7dab1865e0ae39832f8f795.tar.gz
```

[^1]: at least not without further refactoring

Reply:
Fixed that